### PR TITLE
feat: api setup

### PR DESCRIPTION
**Pull request by aburd**
Some refactoring and some boilerplate for getting api functions setup

- api returns a 404 when the user is not logged in
- api formats at utils/response/api
- fixes a bug where an incorrect password through error and brought the server down lol

### CODE REVIEW

File: routes/login.tsx [hunk 1]
```diff
@@ -36,7 +37,12 @@ export const handler = {
       const user = findUserByEmail(db, data.email);
 
       if (!user || !(await checkPassword(user.passwordHash, data.password))) {
-        throw new Error("invalid login");
+        return ctx.render({
```
Comment (aburd) [inline routes/login.tsx:40]: fixes bug, this should not throw

File: utils/response/api.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  msg: string;
+};
+
+export type ApiResponse<T> = {
+  ok: true;
+  data: T;
+} | {
+  ok: false;
+  error: ApiError;
+};
```
Comment (aburd) [inline utils/response/api.ts:16]: - api will give a top level reserved for metadata
- actual data is at data
- if OK is false, then error must be populated
- each error has a key associated with it which can be switched upon to provide control flow and more error information, for now, we just return a msg to provide the user

**Comment by scootyboots** [comment]
looks goooood

i set up a button in my local env to hit `app/api/joke` and it returns the joke or error response as expected

**Comment by aburd** [comment]
Merged~

**Pull request merged**